### DISCUSSION TITLE
Is geting all UNIQUE constraint violated column names possible?

### DISCUSSION THREAD

**Discussion by capymind** [sqlalchemy/sqlalchemy#10649] [Usage Questions] [answered]
Hi, first of all, thanks for reading this dumb(-ish) question.

The point is that I want to get information such as column names
when UNIQUE constraint faild. The example is as follows:

```python
from sqlalchemy import create_engine
from sqlalchemy import MetaData
from sqlalchemy import Table, Column, String
from sqlalchemy import insert

engine = create_engine("sqlite://")
metadata = MetaData()

user_table = Table(
    "users",
    metadata,
    Column("username", String, unique=True),
    Column("nickname", String, unique=True),
)

metadata.create_all(bind=engine)

try:
    with engine.begin() as db:
        db.execute(
            insert(user_table),
            [
                {"username": "sqlalchemy", "nickname": "awesome"},
                {"username": "sqlalchemy", "nickname": "awesome"}
            ]
        )
except Exception as e:
    print(e)
```

with resut below

```text
(sqlite3.IntegrityError) UNIQUE constraint failed: users.nickname
[SQL: INSERT INTO users (username, nickname) VALUES (?, ?)]
[parameters: [('sqlalchemy', 'awesome'), ('sqlalchemy', 'awesome')]]
(Background on this error at: https://sqlalche.me/e/20/gkpj)
```

1. In addition to just printing out the error message, is there a way to get violated column name? such as `e.get_violated_columns()`
2. The above result say `users.nickname` is violated. However, not only `users.nickname` is violated but also `users.username` too. (of course, when I changed nickname's value, the result with `UNIQUE constraint failed: users.username`) I want to __all__ violated columns. Is it possible?

Thank you.

**Comment by zzzeek** [comment, marked as answer]
hi -

for constraint violations, the only facility that exists is to search the error message itself.  it comes straight from the database driver / database server and that's all that SQLAlchemy gets.

for an example of parsing this see what we did in openstack here:

https://github.com/openstack/oslo.db/blob/48ff2a0e43f02dd7f0251391949ca257b96184af/oslo_db/sqlalchemy/exc_filters.py#L160-L192

**Reply by capymind** [reply]
It's a little sad for me, I understand what you have said.
Thanks for the quick and concise explanation, helping me a lot!!

**Comment by auxsvr** [comment]
Some code to achieve this with PostgreSQL:

```python
async def get_constraint_names(schema: str, name: str, pk: bool = False) -> dict[str, list[str]]:
    """Get the name and column names of a PK constraint if `pk` is set, else of the uniqueness
    constraints. The uniqueness constraint names must be retrieved by introspection. Occasionally,
    the primary key name is available without introspection.
    """
    result = {}
    method = "get_pk_constraint" if pk else "get_unique_constraints"
    async with db_session() as session:
        con = await session.connection()
        r = await con.run_sync(lambda c: getattr(sqlalchemy.inspect(c), method)(name, schema))
    if pk:
        r = [r]
    for constraint in r:
        result[constraint["name"]] = constraint["constrained_columns" if pk else "column_names"]
    return result

```
The constraint name is in `e.orig.diag.constraint_name`.

**Reply by capymind** [reply]
Thank you a lot. It helps a lot for me to learn about exceptions and related attribute :)
